### PR TITLE
Add persistent warnings and warnings API

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins static analysis failing

**Comment by monstermunchkin** [comment]
I had to remove the mapped db functions as the generator doesn't support `LEFT JOIN` which is needed for the warnings. Supporting it in the generator would have been trivial but dealing with null strings etc. probably would have been a bit more work.

File: shared/api/warning.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package api
+
+import "time"
+
+// A Warning represents a warning.
```
Comment (stgraber) [inline shared/api/warning.go:5]: Just `Warning represents a warning entry` should do here, no punctuation at the end.

Struct needs to be labeled for swagger.

File: lxd/warnings.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	Delete: APIEndpointAction{Handler: warningsDelete, AccessHandler: allowAuthenticated},
+}
+
+var warningCmd = APIEndpoint{
+	Path: "warnings/{id}",
+
+	Get:    APIEndpointAction{Handler: warningGet, AccessHandler: allowAuthenticated},
+	Put:    APIEndpointAction{Handler: warningPut, AccessHandler: allowAuthenticated},
+	Delete: APIEndpointAction{Handler: warningDelete, AccessHandler: allowAuthenticated},
+}
```
Comment (stgraber) [inline lxd/warnings.go:37]: All of those will need swagger comments on the functions below.

Comment (stgraber) [reply lxd/warnings.go:37]: Should drop the AccessHandler so that only admins can use the API.

File: lxd/api_internal.go [hunk 3]
```diff
@@ -123,11 +126,64 @@ var internalImageOptimizeCmd = APIEndpoint{
 	Post: APIEndpointAction{Handler: internalOptimizeImage},
 }
 
+var internalWarningCreateCmd = APIEndpoint{
+	Path: "warnings",
```
Comment (stgraber) [inline lxd/api_internal.go:130]: Let's make that `testing/warnings`

Comment (stgraber) [reply lxd/api_internal.go:130]: And change image-refresh to match too.

File: client/interfaces.go [hunk 4]
```diff
@@ -314,6 +314,14 @@ type InstanceServer interface {
 	UpdateClusterMember(name string, member api.ClusterMemberPut, ETag string) (err error)
 	RenameClusterMember(name string, member api.ClusterMemberPost) (err error)
 
+	// Warning functions
+	GetWarnings() ([]api.Warning, error)
+	GetWarning(UUID string) (*api.Warning, error)
```
Comment (stgraber) [inline client/interfaces.go:319]: Return ETag.

File: client/interfaces.go [hunk 5]
```diff
@@ -314,6 +314,14 @@ type InstanceServer interface {
...
+	AcknowledgeAllWarnings() error
+	AcknowledgeWarning(UUID string) error
+	DeleteAllWarnings() error
```
Comment (stgraber) [inline client/interfaces.go:322]: Drop all of those and replace with just a single:
`UpdateWarning(UUID string, api.WarningPut, ETag string) error`

File: lxc/warning.go [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+Column shorthand chars:
+
+    c - Count
+    l - Last seen
+    L - Location
+    f - First seen
+    p - Project
+    s - Severity
+	S - Status
```
Comment (stgraber) [inline lxc/warning.go:92]: weird indent here

File: lxc/warning.go [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+
+func (c *cmdWarningDelete) Command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = usage("delete", i18n.G("[<remote>:]<warning-uuid>"))
+	cmd.Aliases = []string{"rm"}
+	cmd.Short = i18n.G("Delete warning")
+	cmd.Long = cli.FormatSection(i18n.G("Description"), i18n.G(
+		`Delete warning`))
+
+	cmd.Flags().BoolVarP(&c.flagAll, "all", "a", false, i18n.G("Delete all warnings")+"``")
```
Comment (stgraber) [inline lxc/warning.go:350]: Let's drop for now, if we get requests, we'll consider the bulk API.

File: lxc/warning.go [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+func (c *cmdWarningAcknowledge) Command() *cobra.Command {
+	cmd := &cobra.Command{}
+	cmd.Use = usage("acknowledge", i18n.G("[<remote>:]<warning-uuid>"))
+	cmd.Aliases = []string{"ack"}
+	cmd.Short = i18n.G("Acknowledge warning")
+	cmd.Long = cli.FormatSection(i18n.G("Description"), i18n.G(
+		`Acknowledge warning`))
+
+	cmd.Flags().BoolVarP(&c.flagAll, "all", "a", false, i18n.G("Acknowledge all warnings")+"``")
```
Comment (stgraber) [inline lxc/warning.go:258]: Let's drop for now, if we get requests, we'll consider the bulk API.

**Comment by stgraber** [comment]
Need to add:
 - i18n
 - doc
 - api extension

**Comment by stgraber** [comment]
@monstermunchkin the new warning test is failing

**Comment by stgraber** [comment]
@tomponline can you do a quick review of this branch in the morning?

**Comment by stgraber** [comment]
@tomponline can you do a quick review of this branch in the morning?

**Comment by monstermunchkin** [comment]
Static analysis is failing because the functions `resolveWarningByUUID` and `resolveWarningByEntityTypeAndEntityID` are currently  not used. They will however be used in the next PR.

**Comment by monstermunchkin** [comment]
I'll remove those two functions and add them to the next PR.

File: lxd/db/entity.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	"github.com/pkg/errors"
+
+	"github.com/lxc/lxd/lxd/db/cluster"
+)
+
+// ErrUnknownEntityID describes the unknown entity ID error
+var ErrUnknownEntityID = fmt.Errorf("Unknown entity ID")
+
+// GetURIFromEntity returns the URI for the given entity type and entity ID.
+func (c *Cluster) GetURIFromEntity(entityType int, entityID int) (string, error) {
```
Comment (tomponline) [inline lxd/db/entity.go:19]: @monstermunchkin in https://github.com/lxc/lxd/pull/8582/commits/2f30dbdba7a151e7d54be5a67f683c46192bb981 you added functions to get objects via their IDs, but this function seems to (mostly) get all objects of a certain type and iterate through them looking for the matching ID. This seems pretty inefficient, especially for instances which will retrieve multiple different queries per instance. Any reason you didnt use the ID oriented retrieval functions you added?

File: lxd/db/warnings.go [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+DELETE FROM warnings WHERE status = ?
+`)
+
+var warningID = cluster.RegisterStmt(`
+SELECT warnings.id FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id LEFT JOIN projects ON warnings.project_id = projects.id
+  WHERE warnings.uuid = ?
+`)
+
+// CreateOrUpdateWarning creates or updates a warning.
+func (c *Cluster) CreateOrUpdateWarning(nodeName string, projectName string, entityTypeCode int, entityID int, typeCode WarningType, message string) error {
```
Comment (tomponline) [inline lxd/db/warnings.go:53]: I've seen the use of "upsert" in the code base for this type of "create or update" action, perhaps "UpsertWarning" would be more consistent with that?

File: lxd/db/warnings.go [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+// GetWarnings returns all available warnings.
+func (c *ClusterTx) GetWarnings() ([]Warning, error) {
+	q := `SELECT warnings.id, IFNULL(nodes.name, "") AS node, IFNULL(projects.name, "") AS project, IFNULL(warnings.entity_type_code, -1), IFNULL(warnings.entity_id, -1), warnings.uuid, warnings.type_code, warnings.status, warnings.first_seen_date, warnings.last_seen_date, warnings.last_message, warnings.count
+	FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id LEFT JOIN projects ON warnings.project_id = projects.id ORDER BY warnings.last_seen_date`
+
+	return c.doGetWarnings(q)
+}
+
+// CreateWarning adds a new warning to the database.
+func (c *ClusterTx) CreateWarning(object Warning) (int64, error) {
```
Comment (tomponline) [inline lxd/db/warnings.go:205]: @monstermunchkin it seems like this function is redundant given we have `CreateOrUpdateWarning` above?

Also, could `CreateOrUpdateWarning` accept the `Warning` struct type argument, rather than individual arguments for clarity?

File: lxd/db/warnings_types.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+// +build linux,cgo,!agent
+
+package db
+
+// WarningType is a numeric code indentifying the type of warning.
+type WarningType int
+
+// Possible values for WarningType
+//
+// WARNING: The type codes are stored in the database, so this list of
```
Comment (tomponline) [inline lxd/db/warnings_types.go:11]: Nice comment :)

File: lxd/db/warnings_types.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	return "Unspecific warning"
+}
+
+// Severity returns the severity of the warning type.
+func (t WarningType) Severity() string {
+	switch t {
+	case WarningClusterTimeSkew:
+		return "low"
```
Comment (tomponline) [inline lxd/db/warnings_types.go:59]: @monstermunchkin feels like severity should be its own Go type, along with constants for the different levels so that the return value of this function can be compared to the discrete severity levels we define?

File: lxd/db/entity.go (hunk 9 above)
Comment (monstermunchkin) [reply lxd/db/entity.go:19]: Previously, I had added `ID` to the filters, and updated the mapper functions (where possible) to handle that. However, that caused issues in the object retrieval and broke our unit tests. I therefore removed the mapper changes, and moved the filtering to the function itself.

File: shared/api/warning.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	// Example: Couldn't find the CGroup blkio.weight, disk priority will be ignored
+	LastMessage string `json:"last_message" yaml:"last_message"`
+
+	// The severity of this warning
+	// Example: low
+	Severity string `json:"severity" yaml:"severity"`
+
+	// The entity affected by this warning
+	// Example: /1.0/instances/c1?project=default
+	Entity string `json:"entity" yaml:"entity"`
```
Comment (tomponline) [inline shared/api/warning.go:53]: @stgraber @monstermunchkin just a thought here, but if `Entity` is always going to be a URL should be call this `EntityURL` perhaps?

File: lxd/db/warnings.go (hunk 10 above)
Comment (monstermunchkin) [reply lxd/db/warnings.go:53]: `SetWarning` seems wrong as we are often creating warnings. But `upsert` might be a good choice.

File: lxd/warnings.go [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+		}
+	}
+
+	// Parse the project field
+	projectName := queryParam(r, "project")
+
+	var dbWarnings []db.Warning
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+
```
Comment (tomponline) [inline lxd/warnings.go:166]: Please remove empty line.

File: lxd/warnings.go [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+		warnings[i] = warning
+	}
+
+	if recursion == 0 {
+		var resultList []string
+
+		for _, w := range filterWarnings(warnings, clauses) {
+			url := fmt.Sprintf("/%s/warnings/%s", version.APIVersion, w.UUID)
+			resultList = append(resultList, url)
+
```
Comment (tomponline) [inline lxd/warnings.go:199]: empty line

File: lxd/warnings.go [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	return response.EmptySyncResponse
+}
+
+func dbWarningToAPIWarning(d *Daemon, warning db.Warning) (api.Warning, error) {
```
Comment (tomponline) [inline lxd/warnings.go:364]: @monstermunchkin this feels like it should be a `ToAPI()` function on the `db.Warning` type (there's previous historic precedent for that with NodeInfo and Certificate).

File: lxd/warnings.go [hunk 18]
```diff
@@ -380,3 +385,37 @@ func dbWarningToAPIWarning(d *Daemon, warning db.Warning) (api.Warning, error) {
...
+		if err != nil {
+			logger.Error("Failed to prune resolved warnings", log.Ctx{"err": err})
+		}
+		logger.Info("Done pruning resolved warnings")
+	}
+
+	return f, task.Daily()
+}
+
+func pruneResolvedWarnings(ctx context.Context, d *Daemon) error {
```
Comment (tomponline) [inline lxd/warnings.go:412]: @monstermunchkin this function doesn't seem correct given the spec says "Automatically prune resolved warnings after 24h", as if the task (for instance) runs daily at 23:59, then a task that has just been resolved at 23:58 would get removed, even though it has been resolved for <24 hours.

File: lxd/api_internal.go [hunk 19]
```diff
@@ -123,11 +126,64 @@ var internalImageOptimizeCmd = APIEndpoint{
...
+type internalWarningCreatePost struct {
+	Location       string `json:"location" yaml:"location"`
+	Project        string `json:"project" yaml:"project"`
+	EntityTypeCode int    `json:"entity_type_code" yaml:"entity_type_code"`
+	EntityID       int    `json:"entity_id" yaml:"entity_id"`
+	TypeCode       int    `json:"type_code" yaml:"type_code"`
+	Message        string `json:"message" yaml:"message"`
+}
+
+func internalCreateWarning(d *Daemon, r *http.Request) response.Response {
```
Comment (tomponline) [inline lxd/api_internal.go:149]: @monstermunchkin as this is a testing route, is it used to artificially create a warning? Would be good to have a comment here to that effect for future clarity.

**Comment by tomponline** [comment]
Overall looks good, no major issues, my main issue is the inefficiency of `GetURIFromEntity`, but thats not a blocker as can be optimised later without changing the interface.

File: lxd/db/warnings.go (hunk 11 above)
Comment (monstermunchkin) [reply lxd/db/warnings.go:205]: > @monstermunchkin it seems like this function is redundant given we have CreateOrUpdateWarning above?

I'd prefer keeping those functions separate due to their length. Merging them would result in a rather long function IMHO.

> Also, could CreateOrUpdateWarning accept the Warning struct type argument, rather than individual arguments for clarity?

That was intentional because a lot of the values in the struct (e.g. ID, UUID, Count, first seen, last seen) would simply be ignored when creating the warning. I thought that approach was clearer.

File: lxd/warnings.go (hunk 18 above)
Comment (monstermunchkin) [reply lxd/warnings.go:385]: No, you're right. The question now is, should we keep resolved warnings for 24h (and then remove them) or clean up resolved warnings every 24h. The latter would be easier, as we'd otherwise need to keep track of the time the warning was resolved.

File: shared/api/warning.go (hunk 14 above)
Comment (stgraber) [reply shared/api/warning.go:53]: Yeah, I think that'd make sense

File: lxd/warnings.go (hunk 18 above)
Comment (stgraber) [reply lxd/warnings.go:385]: We should keep them for 24h, so probably run a task every minute and have a updated_at field in the DB so we can tell how long it's been in that state.

Comment (monstermunchkin) [reply lxd/warnings.go:385]: Could we rather name it `resolved_at` as that's what we're really interested in?

Comment (monstermunchkin) [reply lxd/warnings.go:385]: Actually, to be consistent, it should be `resolved_time` (or `updated_time`).

**Comment by monstermunchkin** [comment]
I was just looking at the CGroup warnings, and believe we should perhaps include the message (or some sort of message) in the db index. Here's the problematic code:
```go
// ...

if !info.Supports(Blkio, nil) {
	msg := " - Couldn't find the CGroup blkio, disk I/O limits will be ignored"
	logger.Warnf(msg)
	warnings.CreateWarning("", "", -1, -1, db.WarningMissingCgroups, msg)
}

if !info.Supports(BlkioWeight, nil) {
	msg := " - Couldn't find the CGroup blkio.weight, disk priority will be ignored"
	warnings.CreateWarning("", "", -1, -1, db.WarningMissingCgroups, msg)
	logger.Warnf(msg)
}

/ ...
```

The result would be a single db entry with only the last warning message, as the unique index is made up of the node, project, entityType, and entityID. We'd only ever know of one missing CGroup whereas the log would show all warnings.

@tomponline @stgraber Any thoughts?

**Comment by stgraber** [comment]
I think we need a separate warning id for each of the cgroup warnings. That way it's possible to cleanly acknowledge one cgroup controller without affecting the others.

So the type would tell us what controller it is and the message would be used to tell us what's missing exactly (probably the path to the missing directory or file).

**Comment by monstermunchkin** [comment]
OK, that would solve the issue.

**Comment by monstermunchkin** [comment]
@stgraber good to go.

File: lxd/db/operations.go [hunk 20]
```diff
@@ -56,7 +73,7 @@ func (c *ClusterTx) GetOperationByUUID(uuid string) (Operation, error) {
 	case 1:
 		return operations[0], nil
 	default:
-		return null, fmt.Errorf("more than one node matches")
+		return null, fmt.Errorf("more than one operation matches")
```
Comment (stgraber) [inline lxd/db/operations.go:76]: Split out as own commit

File: lxd/warnings.go [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+	Path: "warnings",
+
+	Get: APIEndpointAction{Handler: warningsGet},
+}
+
+var warningCmd = APIEndpoint{
+	Path: "warnings/{id}",
+
+	Get:    APIEndpointAction{Handler: warningGet},
+	Put:    APIEndpointAction{Handler: warningPut},
```
Comment (stgraber) [inline lxd/warnings.go:35]: We're missing patch here, should be made to be identical to Put for now.

File: client/interfaces.go [hunk 22]
```diff
@@ -314,6 +314,12 @@ type InstanceServer interface {
 	UpdateClusterMember(name string, member api.ClusterMemberPut, ETag string) (err error)
 	RenameClusterMember(name string, member api.ClusterMemberPost) (err error)
 
+	// Warning functions
+	GetWarnings() (warnings []api.Warning, err error)
```
Comment (stgraber) [inline client/interfaces.go:318]: Need GetWarningUUIDs() for consistency

**Comment by stgraber** [comment]
Fixed the few things I noted above.

**Comment by stgraber** [comment]
Ah, we also need the API in the old `rest-api.md`, adding that now.

**Comment by stgraber** [comment]
Swagger doc also needed to be regenerated (`make update-api`).
